Dockerization

Freddo3000 opened the pull request:
This pull request dockerizes the project, making it easier to spool up and host. For this a new Dockerfile is added, which I've tried to keep lightweight using Alpine. Two sample `docker-compose.yml`s are also provided, one running simple sqlite and the other running postgresql. To make it more production-safe, the docker version runs with the [gunicorn](https://gunicorn.org/) WSGI server. 

Some additional changes to the repo is that I've moved the source into a separate `src` directory, to simplify the docker build. I've also edited `config.py` to support running using only environment variables, to ease docker configuration. It also adds a workflow for publishing the docker image to https://ghcr.io, which can be previewed on my repo [here](https://github.com/Freddo3000/gnucash-web/packages).

edit:

resolves #44, as that was an issue I also noticed.

joshuabach commented:
Hey Freddo,

thanks for the PR, Dockerisation would indeed be a good addition. But boy it is a big one ^^ It'll take some time to go through this. 

A few things I have spotted already:

## Superfluous files 
I noticed that you commited your `.idea` folder, which I think is IDE-specific metadata that should not be included in the repo. It also seems that the `src/` subfolder contains build files. At least some of these are listed in `.gitignore`, but only if they appear on the top level. Can you remove those, update `.gitignore` and check that there are no other files that you accidentally commited? 
 
## Commit structure
I think your commits are a bit scrambeled. I'm not a fan of squashing PRs into a single commit, especially if they do a few separate things. On the other hand, at least your "tweak docker compose" hotfix-commit could be squashed into the first commit (nitpick: which should be named "Add docker setup", or sth. similar, see [subject naming conventions](https://github.com/naming-convention/naming-convention-guides/blob/master/git/commit-message-naming.md#subject)). Idealy, I think there should be these commits:
1. Move files to src/ subdirectory
2. Add support for using environment variables only for configuration
3. Add docker setup 

Number 3 could include the github action and the README update since those are both inherently part of an dependent on the dockerisation.

That would make reviewing much easier. However, at least the second part requires some git knowledge to do. If you have any more questions or need help, just say so.

Thank you!

Freddo3000 commented:
I've removed the .idea directory and added it to .gitignore, and the accidentally committed build/dist files (I was wondering why Git wasn't detecting moved files...). As for commit structure, I was as you mentioned working on the assumption that the entire PR would be squashed and merged, so I didn't pay close attention to it. If  you'd like, I could redo the changes in a new PR with a git history like you describe.

joshuabach commented:
Without the accidentally commited files, the PR looks much more managable, so you can leave it as it is and I will squash it upon merge. I hope I can review this on Friday.

joshuabach commented on README.md:
```diff
@@ -130,6 +130,14 @@ chmod-socket = 660
+
+*GnuCash Web* can be run using [Docker](https://www.docker.com/), either using the published 
+[packages](https://github.com/joshuabach/gnucash-web/packages), DockerFile, or using [docker compose](https://docs.docker.com/compose/) 
+with the provided sample `docker-compose.yml` files.
```
Can you add a brief description on how the database works when using the Docker setup? Namely,
1. That its a postgres DB with passthrough auth
2. How to configure username+password
3. How to access the database directly (e.g. using the GnuCash Desktop client)
4. If the section "Initialising database" can be applied 1:1, and if not, what has to be done differently

And anything else that comes to mind.

How to configure gnucash-web generally when using docker would also be helpful

Freddo3000 replied:
Do you want this in the readme or in the docker-compose file? I'd argue that the rest of the documentation is sufficient to understand the docker-compose configuration, unless you want to explain how the Docker Compose plugin itself works alongside the Postgres image. For the general Dockerfile, I could add a table describing the possible environment variables if that is desired. 

There shouldn't be any difference compared to regular setup other than what's required to run and connect to Docker, which depends on the individual setup and network/firewall configuration.

joshuabach replied:
I mean in the README. This does not have to be verbose, just some pointers on what needs to be done apart from running `docker-compose up`.

For example, someone not familiar with docker might struggle on how to connect to the database directly, since it does not map a port. This should be briefly explained.

joshuabach replied:
I mean, it can be as simple as:

---

*GnuCash Web* can be run using [Docker](https://www.docker.com/), either using the published 
[packages](https://github.com/joshuabach/gnucash-web/packages), DockerFile, or using [docker compose](https://docs.docker.com/compose/) 
with the provided sample `docker-compose.yml` file. This file uses a simple SQLite backend. A file `docker-compose-postgres.yml' with a PostgreSQL backend is also provided. 

In either case, configuration is done via environment variables in the compose file instead of the default configuration file. The same options are available.

To access the postgres database, expose the postgres port `5432` on a host port (think about security before doing that) in the compose file [add link to explanation in docker docs] or connect to the container on that port directly.

Note that while the SQLite version includes a sample database, the PostgreSQL is uninitialized, see [below]() [add link] for more information.

---

Or something similar in that spirit.

joshuabach commented on src/bootstrap:
While moving the submodules, you seem to have also updated them (e.g. this one changed from 60714e to  101ff7). This applies to all the submodules as far as I can see.

Hiding this in this change (and even while simultaneously moving the submodules) makes it hard to track this down in case in the future an issue is caused by one of these updates, so I'd prefer keeping this separate.

Can you revert the updates, unless specifically required for this change?

joshuabach commented on sample/sample.sqlite:
What did you change in the sample file?

Freddo3000 replied:
No changes, new file. It is simply an default settings gnucash sqlite database used in the sqlite example docker compose.

joshuabach replied:
Ah, I guess i mixed this up with the sample file running on [the demo](https://gnucash-web-demo.bachmeier.cc). I think i never commited that file.

joshuabach commented on src/requirements.txt:
```diff
@@ -1,6 +1,6 @@
 mysql==0.0.3
 piecash==1.2.0
-pycryptodome==3.12.0
+pycryptodome==3.18.0
```
Is this update required for your changes?

Freddo3000 replied:
I encountered issues building wheels using the older version, as up until 3.16 it didn't support musl Linux which Alpine uses.

joshuabach commented on Dockerfile:
```diff
@@ -0,0 +1,40 @@
+
+EXPOSE 8000
+
+ENV SECRET_KEY='0123456789ABCDEF0123456789ABCDEF'
```
maybe change this to '000000....0000' or 'INSECURE' or something similar and add a comment to make it more apparent that this should not be used in production like this

joshuabach commented on Dockerfile:
```diff
@@ -0,0 +1,40 @@
+      libpq-dev
+
+COPY ./src/requirements.txt .
+RUN pip wheel --no-cache-dir --wheel-dir /wheels mysql pycryptodome psycopg2
```
Why not -r requirements.txt? You even copy the file into the container

Freddo3000 replied:
`mysql`, `pycrptodome`, and `psycopg2` need additional dependencies for compilation, so it is built in a separate stage. The requirements file can likely be moved from this stage to the next one, where it is used in 
`pip install --no-cache /wheels/* .[pgsql,mysql]`.

joshuabach commented on docker-compose-postgres.yml:
Maybe name this file "docker-compose-with-postgres" to make clear that its not just a compose file for postgres, but for gnucash-web with postgres?

Freddo3000 replied:
Having it simply be `postgres` makes the name easier to type out and autocomplete, as this type of configuration is more often than not done by terminal. It isn't unreasonable to assume a user that wants to use Postgres will inspect the postgres compose file.

joshuabach commented on .github/workflows/docker-publish.yml:
```diff
@@ -0,0 +1,100 @@
+          registry: ${{ env.REGISTRY }}
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      # Extract metadata (tags, labels) for Docker
+      # https://github.com/docker/metadata-action
+      - name: Extract Docker metadata
+        id: meta
+        uses: docker/metadata-action@98669ae865ea3cffbcbaa878cf57c20bbf1c6c38
+        with:
+          images: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}
+
+      # Build and push Docker image with Buildx (don't push on PR)
+      # https://github.com/docker/build-push-action
+      - name: Build and push Docker image
+        id: build-and-push
+        uses: docker/build-push-action@ac9327eae2b366085ac7f6a2d02df8aa8ead720a
+        with:
+          context: .
+          push: ${{ github.event_name != 'pull_request' }}
+          tags: ${{ steps.meta.outputs.tags }}
+          labels: ${{ steps.meta.outputs.labels }}
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+
+
+      # Sign the resulting Docker image digest except on PRs.
+      # This will only write to the public Rekor transparency log when the Docker
+      # repository is public to avoid leaking data.  If you would like to publish
+      # transparency data even for private images, pass --force to cosign below.
+      # https://github.com/sigstore/cosign
+      - name: Sign the published Docker image
+        if: ${{ github.event_name != 'pull_request' }}
+        env:
+          # https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-an-intermediate-environment-variable
+          TAGS: ${{ steps.meta.outputs.tags }}
+          DIGEST: ${{ steps.build-and-push.outputs.digest }}
+        # This step uses the identity token to provision an ephemeral certificate
+        # against the sigstore community Fulcio instance.
+        run: echo "${TAGS}" | xargs -I {} cosign sign --yes {}@${DIGEST}
```
This seems very verbose and I guess comes literally from the sample file. Do we need all of it?

Freddo3000 replied:
This is the suggested Docker image publish action by github, only change being to recursively pull submodules. I can't find a direct link to it, but you can see a description of it here: https://docs.github.com/en/actions/publishing-packages/publishing-docker-images

It can be cut down when it comes to comments, but I wanted to stay close to the template.

joshuabach replied:
Better to leave the comments in, might be helpful sometime, esp. if we don't know what all of this does :-)

joshuabach reviewed (changes requested):
I left a few questions / suggestions.

Note that this is a "cold" review. I still have to try it out.

joshuabach commented:
So functionality-wise this seems to be working great, thanks again!

Only the README-comment open then. This is not strictly a blocker for merge, but while we are at it, i would prefer getting it right.

Please note that i am unavailable from monday until the 23rd, so I can't guarantee we can get this merged before that. But it will definitly make it into the next release.

joshuabach commented on README.md:
Maybe also add a note to the config section, that config variables can also be set by env variable

Freddo3000 commented:
Finally sat down working on this after being swamped for a month...

joshuabach commented:
Ok now I guess it was my turn to not respond you in a while :-)

Anyway looks good. Thank you again for your contribution and your effort of interacting with me!
